### PR TITLE
ci: run workflows on changed components instead of every pull request

### PR DESCRIPTION
**Pull request by kakkoyun**
## What

Makes CI run on what a change can actually affect, instead of running everything on every pull request.

## Why

PR #5346 changed four files, all under `openfeature/`. It produced **204 check runs and ~623 aggregate job-minutes**:

| Workflow | Jobs | Job-min |
|---|---|---|
| System Tests | 120 | 320 |
| Pull Request Tests | 29 | 131 |
| Orchestrion | 12 | 82 |
| Check PR CI Status (polling) | 1 | 19 |
| Govulncheck | 3 | 17 |
| Static Checks | 10 | 16 |
| Parametric / CodeQL / Generate | 3 | 11 |

The merge queue then re-ran most of it on `mq-working-branch-main-*`: another **178 jobs / 562 job-min**. So a merged PR costs roughly **1,185 job-minutes** today.

Classifying the last 80 merged PRs by what they touch: 49% are core changes that legitimately need everything, but 16% are CI-config-only, 11% touch only a leaf package (`openfeature`, `profiler`, `civisibility`, `crashtracker`), and 9% are contrib-only. That is about a third of PRs paying full price for work that cannot be affected.

## The design problem

Native `paths:` filters are an allowlist, and allowlists are unsafe by default: a brand-new directory matches nothing, so its workflow silently stops running. Nobody notices until something ships broken.

So this inverts the default. [`.github/ci-components.yml`](.github/ci-components.yml) maps every path to the CI work it requires, and **a path matching no component enables every gate**. The table is an allowlist of paths that are *provably safe to skip*; being absent from it costs money, never coverage. `make lint/misc` fails on any unclassified path or unnamed top-level entry — the same no-catch-all discipline as `CODEOWNERS`.

## Why directory globs were not enough

Measured with `go list -deps` across all 85 workspace modules. Several intuitions turned out to be wrong:

- **Every contrib transitively imports 90–100 root-module packages** (gin alone: 95). A `ddtrace/tracer` change really does need all ~70 of them. There is no pruning to be had there, and pretending otherwise would be the bug.
- **Fan-out varies wildly inside one directory.** `datastreams/options` is reachable from 77 modules; the rest of `datastreams/` from 8. `internal/orchestrion` from 78, but `internal/orchestrion/_integration` from none. Directory-level rules are the wrong granularity.
- **`contrib/os/` has no `go.mod` at any depth.** It is root-module code wearing a `contrib/` path. A `contrib/**` rule would have narrowed a core change.
- **Three gated GitLab microbenchmarks live in contrib modules** (`BenchmarkGraphQL`, `BenchmarkHttpServeTrace`, `BenchmarkStartSpan`), so "contrib changes cannot affect benchmarks" would also have been wrong.

Each of these is now a test case rather than a comment.

## Why skipping is safe here

- `gh api repos/DataDog/dd-trace-go/rulesets` shows the only required status check on `main` is `devflow/mergegate`. No Actions job name is registered, and classic branch protection returns 404.
- `all-green.yml` uses `DataDog/ensure-ci-success`, which discovers checks dynamically and counts `skipped`/`neutral` as success.
- **Demonstrated, not assumed:** #5346 merged with API Semantic Diff, Build Metrics, config-audit and Dynamic Analysis Checks producing *zero* check runs. Seven workflows already rely on this.

No "skip-but-mark-done" plugin is needed.

## What changed

- **New** `.github/ci-components.yml`, `scripts/ciselect/`, `.github/actions/changed-components/`.
- **Gated** `pull-request`, `system-tests`, `orchestrion`, `static-checks` (per job), `parametric-tests`, `codeql-analysis`, `generate`.
- **Retightened** `dynamic-checks` (its `**/go.mod` fired two full test matrices on every Dependabot contrib bump — 85 job-min) and `govulncheck` (its `**/*.go` matched every PR).
- **Narrowed** the contrib matrix from a fixed 6-way round-robin over all modules to a module-graph reverse-dependency closure.
- **Scoped** the GitLab microbenchmarks, which ran 45-minute jobs on 48-core metal runners on every branch.

Join jobs needed rework: they tested `result != 'success'`, which reads a gated-off `skipped` dependency as a failure. They now allowlist the failing results, and each carries an explicit `needs.changes.result != 'success'` clause — without it a crashed classifier leaves everything downstream `skipped` and the gate reports green having run nothing.

Every gate is written `github.event_name != 'pull_request' || ...`, so push, schedule, `workflow_call` and the merge queue are untouched by construction rather than by trusting the classifier to fail open.

## Guards

| Guard | What it catches |
|---|---|
| `coverage_test.go` | An unclassified path or a new top-level entry. Wired into `make lint/misc`. |
| `rules_test.go` | Regressions on the cases above, plus fail-open behaviour |
| `workflows_test.go` | Unregistered workflows, undeclared native filters, dead globs, a join job that lost its `changes` dependency |
| `action_test.go` | The composite action's fail-open gate list drifting from the table |
| `gitlab_test.go` | A benchmark whose package the GitLab path list does not cover |
| `depgraph_test.go` | Recorded reverse-dependency facts going stale (nightly, `go list -deps`) |

## Verification

Local: `go build ./...`, `go vet` (both tag sets), `go test -race ./scripts/ciselect/`, `golangci-lint`, `actionlint` (all workflows), `check_copyright`, `check_codeowners`, `gofmt` — all pass. `go test -tags depgraph` passes and independently re-derives the recorded reverse-dependency facts.

### Measured in CI

Four canary pull requests, each a one-line no-op change, all targeting this branch so they diff against it rather than against `main`. **All five PRs are green.**

| Change shape | PR | Jobs | Skipped | Job-min |
|---|---|---|---|---|
| this PR (`.github/actions/**` → `universal`) | #5353 | 319 | 4 | 1514 |
| single contrib (`contrib/gin-gonic/gin/gintrace.go`) | #5356 | 177 | 7 | 733 |
| leaf package (`openfeature/provider.go`) | #5354 | 167 | 13 | 639 |
| CI config only (`.github/workflows/generate.yml`) | #5357 | 41 | 23 | **18** |
| docs only (`README.md`) | #5355 | 41 | 24 | **15** |

**Read these in job counts, not minutes.** System Tests alone is ~117 jobs and swung between 330 and 493 job-minutes across these runs for an identical job set, so minutes are dominated by runner noise. And the 1514 for this PR includes Build Metrics (773) and Dynamic Analysis (85) firing *because this PR edits those workflow files* — correct self-validation, not a typical core PR.

Against the real pre-change baseline (#5346, an openfeature-only PR: 178 real jobs):

| Change shape | Real jobs | vs baseline |
|---|---|---|
| leaf package | 152 | **−15%** |
| single contrib | 170 | −4% |
| CI config only | 18 | **−90%** |
| docs only | 17 | **−90%** |

What actually changed, per shape:

- **docs / CI-config-only**: everything of substance skipped. System Tests 117 jobs → 2. Orchestrion 12 → 2. Static Checks 10 → 3. Total wall cost ~16 job-minutes. The `System Tests`, `Orchestrion Integration Tests` and `PR Unit and Integration Tests / *` checks all still report **green**.
- **leaf package**: Orchestrion (−10 jobs), Govulncheck (−3), Parametric and 12 of the contrib-matrix jobs all drop. System Tests stays, correctly — see the review note below.
- **single contrib**: the contrib matrix narrows from 6 chunks over ~72 modules to exactly two, `./contrib/gin-gonic/gin/` and `./instrumentation/internal/namingschematest/` (a genuine reverse dependency, since it requires every contrib). System Tests and Orchestrion stay, as designed.
- **core / universal**: unchanged.

### Honest note on the headline number

An earlier revision of this PR measured the leaf case at 143 job-minutes. That was wrong: `openfeature` was skipping System Tests, which review correctly identified as a coverage hole (the `FEATURE_FLAGGING_AND_EXPERIMENTATION` scenario is that feature's only end-to-end coverage). With the gate restored the leaf saving is −15%, not −91%.

So the savings concentrate in three places, not four:

1. **Docs and CI-config-only PRs** — ~90% fewer jobs. These were 16%+ of the last 80 merged PRs.
2. **The retightened filters** — `dynamic-checks` was firing two full test matrices (85 job-min) on every Dependabot contrib bump; `build-metrics` was firing 107 jobs / 773 job-min whenever anyone edited `.github/workflows/orchestrion.yml`; `govulncheck` was running on every PR.
3. **The narrowed contrib matrix** — 72 modules down to the affected ones.

**System Tests is now clearly the dominant remaining cost** (~117 jobs, 330–493 job-min) and it runs for core, contrib and leaf changes alike. Per-weblog-variant selection is the highest-value follow-up, and it needs the weblog→contrib mapping that lives in `DataDog/system-tests` rather than here.

### Bugs found by the canaries and by review

Four, all fixed with regression tests:

1. **Empty selection read as "no selection"** (canary). `ciselect` printed nothing when no contrib module needed testing, and the workflow reads it via `${CONTRIB_MODULES:-ALL}` — shell substitutes the default for an *empty* value as well as an unset one, so "test nothing" became "test all 72 modules". Explicit `NONE` sentinel now, with a test asserting the CLI always prints a verdict.
2. **CRLF** (canary). Three tests parse YAML with line-anchored regexes; Windows checkouts carry `\r\n`, so the patterns found nothing and `scripts/ciselect` failed on `windows-latest` — inside both `test-core` and `multios-unit-tests`, since the package is part of the root module. Reads now normalise first; the guard writes a real CRLF file and reads it back, and was confirmed to fail when the normalisation is removed.
3. **Dependency-induced skips passed the summary gates** (review — thanks @darccio). GitHub applies an implicit `success()` to every `needs:`, so a job whose prerequisite failed reports `skipped`, exactly like a deliberately gated-off job. Allowlisting every skip meant a failing `build-haproxy` skipped `system-tests` and the stable check went green having run nothing — strictly worse than the `!= 'success'` test it replaced. Each summary job now requires that the classifier succeeded, that no dependency failed or was cancelled, and that the work really succeeded whenever its gate is on.
4. **CI-config edits did not run the invariants** (review). `make lint/misc` in the `copyright` job was the only caller of `go test ./scripts/ciselect/`, so a workflow-only or `.gitlab/`-only PR never ran `workflows_test.go` or `gitlab_test.go` — precisely the changes they police. Both components now enable `static-actions`, `check-github-actions` runs the tests, and a test asserts that loop stays closed.

Worth noting how (2) surfaced: `PR Unit and Integration Tests / test-core` went red on both the machinery PR and the leaf canary. That was a join job correctly propagating a real Windows failure through a run where most jobs were gated off — evidence the gating does not mask failures.

## For reviewers

1. **CodeQL** — I could not read `repos/.../code-scanning/default-setup` (needs `admin:repo_hook`). Worth confirming GitHub Advanced Security is not configured as an independent merge gate. Note CodeQL still runs unfiltered on `push: main` and `mq-working-branch-**`, so `main` coverage is unchanged either way.
2. **Govulncheck** — the retightening deserves an explicit look. `push` has no paths filter, so every merge-queue branch is scanned in full before landing; nothing reaches `main` unscanned. If security prefers otherwise it is a one-line revert.
3. **Merge queue is deliberately out of scope** — 47% of the spend. The MQ run exists to catch interaction with other concurrently-merging PRs, which is exactly what per-PR file classification does not model. Worth a separate discussion.
6. **The new failure path is not empirically exercised.** The pass path (gated off → green) is covered by the canaries, and `TestJoinJobsRejectDependencyInducedSkips` pins the condition's shape, but no run has deliberately failed a prerequisite to watch a summary gate go red. Happy to force one if you want that proof before merging.
4. `all-green.yml` sets `initial-delay-seconds: "600"`, so a docs-only PR will now finish CI in ~2 minutes and then wait 8 more. Lowering that is a one-line follow-up, without which much of the developer-visible win is invisible.
5. The six commits are independently reviewable and land in dependency order.

## Pre-existing problems found along the way

- **`build-metrics.yml` over-triggers.** Its `**/orchestrion.yml` filter also matches `.github/workflows/orchestrion.yml`, so editing *that workflow* has always fired a **107-job, 765-job-minute** matrix for nothing. Excluded explicitly here, since it is the same category of broken filter as the other two. The component table already resolved that path correctly — non-code components sort ahead of code ones — but nothing asserted it; now a test does.
- **`BenchmarkAgentTraceWriter` does not exist.** It is listed in `microbenchmarks-1`, which feeds `pr-performance-gates`, and the runner matches it with `-bench ^BenchmarkAgentTraceWriter$` — so it has been measuring nothing. Recorded in a `knownMissingBenchmarks` exception rather than fixed here; tracked separately.
- **`ecosystems-label-pr.yml` did not self-reference.** Editing it could not trigger it. Fixed, since the new test requires it.

## Canary pull requests

#5354, #5355, #5356 and #5357 are throwaway drafts targeting this branch, kept open so the measurements above can be inspected. They should be closed before this merges.

### CODE REVIEW

File: .github/workflows/system-tests.yml [hunk 1]
```diff
@@ -436,21 +480,27 @@ jobs:
...
+          needs.changes.result != 'success'
+          || contains(fromJSON('["failure", "cancelled"]'), needs.build-weblog-images.result)
+          || contains(fromJSON('["failure", "cancelled"]'), needs.system-tests.result)
+          || contains(fromJSON('["failure", "cancelled"]'), needs.tracer-release.result)
         run: echo "Failure!" && exit 1
+      - name: Skipped
+        if: needs.system-tests.result == 'skipped'
+        run: echo "No changed path requires system tests (see .github/ci-components.yml)."
+      - name: Success
+        run: echo "Success!"
```
Comment (darccio) [inline .github/workflows/system-tests.yml:520]: The revised `system-tests-done` summary treats `system-tests.result == 'skipped'` as an intentional CI-scope skip. But `system-tests` also becomes skipped if any prerequisite fails, due to GitHub Actions’ implicit `success()` condition.

For example:

1. A changed path enables `system-tests`.
2. `build-haproxy` fails (`:429-445`).
3. `system-tests`, which needs `build-haproxy` (`:155-161`), is skipped.
4. `system-tests-done` does not depend on `build-haproxy` (nor the other image-build jobs or `warm-repo-cache`).
5. Its failure condition only rejects `failure`/`cancelled` from `build-weblog-images`, `system-tests`, and `tracer-release` (`:495-500`). It accepts the skipped `system-tests` result and finishes successfully.

This regresses the previous behavior, which failed the stable **System Tests** check for any non-successful system-test job.

**Suggested fix:**
- Add these dependencies to `system-tests-done.needs`:
  - `warm-repo-cache`
  - `build-service-extensions-callout`
  - `build-haproxy`
  - `build-apim-callout`
- Extend the Failure step’s allowlist check to fail when any of those jobs is `failure` or `cancelled`.
- Preserve acceptance of `skipped` only for the intentionally gated-off case.
- Add a regression test that ensures the stable summary observes failures from every prerequisite that can cause `system-tests` to be skipped.

**Review by darccio — APPROVED** [review]
Only one change required.

File: .github/workflows/system-tests.yml (hunk 1 above)
Comment (kakkoyun) [reply .github/workflows/system-tests.yml:520]: Confirmed and fixed in 6b80266 — you are right that this was a regression, not just an edge case. Added `warm-repo-cache` and all three image-build jobs to `system-tests-done.needs` and to the failure allowlist as you suggested, plus a clause that requires real success whenever the gate is on (`needs.changes.outputs.system-tests == 'true' && needs.system-tests.result != 'success'`) — that is what separates an intentional skip from a collateral one. Same fix applied to `pull-request-tests-done`, `integration-test-done` and the `test-contrib` join, which all had the same hole. `tracer-release` keeps allowlist-only treatment since its skip is schedule-driven and legitimate. `TestJoinJobsRejectDependencyInducedSkips` now pins the shape so it cannot be dropped while editing these conditions.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: .github/workflows/pull-request.yml [hunk 2]
```diff
@@ -57,14 +82,24 @@ jobs:
...
     steps:
+      # Failure is an allowlist, not `!= 'success'`: a gated-off dependency
+      # reports 'skipped', which must not read as a failure. `changes` itself is
+      # checked explicitly -- if the classifier dies, everything downstream is
+      # 'skipped' and this gate would otherwise go green without running a test.
+      - name: Failure
+        if: >-
+          needs.changes.result != 'success'
+          || contains(fromJSON('["failure", "cancelled"]'), needs.unit-integration-tests.result)
+          || contains(fromJSON('["failure", "cancelled"]'), needs.multios-unit-tests.result)
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply .github/workflows/pull-request.yml:106]: Valid — fixed in 6b80266. `warm-repo-cache` is now a dependency of `pull-request-tests-done` and in its failure allowlist, and the gate only accepts a skip when the classifier says the suite was off: `needs.changes.outputs.pull-request-tests == 'true' && (... .result != 'success')`. Applied the same treatment to `system-tests-done`, `integration-test-done` and the `test-contrib` join (keyed on `needs.set-up.outputs.has-contribs`).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: .github/ci-components.yml [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+  # --------------------------------------------------- leaf root-module packages
+  # No contrib imports these. `dependent-modules` is the measured set of
+  # workspace modules that do; depgraph_test.go re-derives it nightly.
+  - id: openfeature
+    paths:
+      - openfeature/**
+      - openfeature/ffe-system-test-data   # git submodule; matched as a bare gitlink
+    dependent-modules: []
+    gates: [ "@go-hygiene", pull-request-tests ]
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply .github/ci-components.yml]: Correct, and this was a real coverage hole — `.github/workflows/system-tests.yml:243` runs `FEATURE_FLAGGING_AND_EXPERIMENTATION`, and `openfeature/ffe-system-test-data` is that scenario's fixture. Added `system-tests` to the component in 6b80266, with the rationale in the `why:` field, and updated the two test cases that had encoded the wrong expectation.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: .github/ci-components.yml [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+  # ----------------------------------------------- not inputs to any Go build
+  # These precede every code component: no file here can change what the Go
+  # toolchain compiles, so claiming them first keeps a doc edit inside a code
+  # directory (contrib/README.md) from dragging in that directory's test suite.
+  # ------------------------------------------------------- CI config and docs
+  - id: workflows
+    paths: [ .github/workflows/*.yml ]
+    self-reference: true
+    gates: [ static-actions ]
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply .github/ci-components.yml:177]: Verified and fixed in 6b80266. `make lint/misc` was indeed the only caller of `go test ./scripts/ciselect/`, and it runs in the `copyright` job gated on `static-copyright`. The `workflows`, `workflow-codegen` and `gitlab` components now enable `static-actions`, and `check-github-actions` runs the tests directly — these validate CI configuration rather than Go code, so that job is their natural home. `TestCIConfigGateRunsTheseTests` asserts both halves so the gate cannot become decorative.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: .github/ci-components.yml [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+    why: scripts/check_codeowners.go runs in make lint/misc, alongside copyright.
+    paths: [ CODEOWNERS ]
+    gates: [ static-copyright ]
+
+  - id: gitlab
+    why: Separate pipeline. .gitlab/ is its own Go module, not imported here.
+    paths:
+      - .gitlab/**
+      - .gitlab-ci.yml
+    gates: []
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply .github/ci-components.yml]: Fixed alongside the workflow case in 6b80266 — `.gitlab/**` and `.gitlab-ci.yml` now enable `static-actions`, which runs `go test ./scripts/ciselect/` in `check-github-actions`, so `TestBenchmarkChangesCoverEveryBenchmark` fires when the `BENCHMARKS` variables or the `.benchmark-changes` list are edited. Worth noting that test already found one such case: `BenchmarkAgentTraceWriter` does not exist and has been measuring nothing, recorded in `knownMissingBenchmarks` and tracked separately.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: .github/workflows/system-tests.yml (hunk 1 above)
Comment (kakkoyun) [reply .github/workflows/system-tests.yml:520]: Followed up with two throwaway canaries to prove the failure path, since only the pass path had been exercised. Both edited just `orchestrion.yml`, so the self-reference rule kept them to 6 and 11 job-minutes.

**Your exact scenario** — prerequisite fails while the gate is on:

| job | conclusion |
|---|---|
| `Detect changed components` | success (so the gate was on) |
| `Service containers` | **failure** (forced) |
| `Integration Test (...)` | skipped |
| **`Orchestrion Integration Tests`** | **failure** ✅ |

And with the classifier itself dead, all four gated jobs reported `skipped` and the summary check still went red — the `needs.changes.result != 'success'` clause.

Both canaries are now closed; the runs remain linked above if you want to inspect them.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by darccio — APPROVED** [review]
LGTM
